url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+                	address = new InetSocketAddress(packet.getAddress(),packet.getPort());           				
+            		sendQueue.put(jsonToDatagram(message,address)); 
+            	}else{ // else forward
+            		message = jsonObj;
+            		message.put("ip_address",NetworkProperties.nAddress);//current's address changes each hop
+            		if(routingTable.containsKey(target_id)){
+            			address = new InetSocketAddress(routingTable.get(target_id).getIp(),NetworkProperties.nPort);
+            			sendQueue.put(jsonToDatagram(message,address));
+            		}else{
+            			
+            		}             	           				           		 
+            	}           	
+                break;
+                //internal housekeeping. send out pings to stale contacts
+            case REFRESH_ROUTING_TABLE:
+            	//System.out.println("will handle update routing table message");
+               	for (Map.Entry<Integer, Contact> entry : routingTable.entrySet()){
+               		final Contact contact = entry.getValue();
+               		long currTime = System.currentTimeMillis();
+               		if(currTime - contact.getTimestamp() > NetworkProperties.staleTime*1000){	               		
+	        			message = getPingJson(contact);
+	        			address = new InetSocketAddress(contact.getIp(),NetworkProperties.nPort);           				
+				        sendQueue.put(jsonToDatagram(message,address));
+				        Timer pingTimer = new Timer();
+				        pingTimer.schedule(new TimerTask() {
+							//if timer not cancelled before time expires then contact removed
+							public void run() {
+								routingTable.remove(contact.getId());							
+							}
+						}, NetworkProperties.pingWait*1000);
+				        pingTimers.put(contact.getId(), pingTimer);
+	        		}
+               	}
+            	break;
+        }
+        }catch(InterruptedException e){
+          System.err.println("Java!! You ain't serious, are you? Anyway, couldn't put message on sendQ in Controller");
+          e.printStackTrace();
+        }
+    }
+
+
+
+	@SuppressWarnings("unchecked")
+    public void sendJoin(InetSocketAddress bootstrap_node,
+                         int nodeHash,
+                         int bootHash){
+        nodeID = nodeHash;
+        bootstrapNode = bootstrap_node;
+        if(nodeHash == bootHash) return;
+
+        JSONObject sendJSON = new JSONObject();
+        sendJSON.put("type",MessageType.JOINING_NETWORK_SIMPLIFIED.name());
+        sendJSON.put("node_id",Integer.toString(nodeHash));
+        sendJSON.put("target_id",Integer.toString(bootHash));
+        sendJSON.put("ip_address",NetworkProperties.nAddress.toString());
+        try {
+            sendQueue.put(jsonToDatagram(sendJSON,bootstrap_node));
+        } catch (InterruptedException e) {
+            System.err.println("Could not put Datagram on sendQueue in Controller");
+            e.printStackTrace();
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 292:297, 308:313, 339:344.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.